cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/DataMigrator.java

```diff
@@ -75,20 +75,36 @@ public void run() {
       try {
         DatasetId v2MetricsTableId = getMetricsDatasetId(v2TableNamePrefix, resolution);
         DatasetId v3MetricsDatasetId = getMetricsDatasetId(v3TableNamePrefix, resolution);
-        if (MigrationTableHelper.hasInstanceWithRetry(datasetFramework, v2MetricsTableId)) {
+        Boolean hasInstance =  MigrationTableHelper.hasInstanceWithRetry(datasetFramework, v2MetricsTableId);
+
+        if (hasInstance == null) {
+          // hasInstance returns null during stop; so safe to break
+          break;
+        }
+
+        if (hasInstance) {
           try (MetricsTable v2MetricsTable =
                  MigrationTableHelper.getDatasetWithRetry(datasetFramework, v2MetricsTableId);
                MetricsTable v3MetricsTable =
                  MigrationTableHelper.getDatasetWithRetry(datasetFramework, v3MetricsDatasetId)) {
+            if (v2MetricsTable == null || v3MetricsTable == null) {
+              // we get null when we are stopping, so safe to break
```

Comment:
this is ambiguous. You also get null if the table does not exist. Returning null when stopping is not the right answer.